https://huggingface.co/docs/text-generation-inference/main/en/reference/launcher#kvcachedtype

      --kv-cache-dtype <KV_CACHE_DTYPE>
          Specify the dtype for the key-value cache. When this option is not provided, the dtype of the model is used (typically `float16` or `bfloat16`). Currently the only supported value are `fp8_e4m3fn` and `fp8_e5m2` on CUDA
          
          [env: KV_CACHE_DTYPE=]
          [possible values: fp8_e4m3fn, fp8_e5m2]


In [1]:
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer

# تحميل النموذج والمحول
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
kv_cache_dtype = torch.float16  # يمكنك استخدام fp8_e4m3fn أو fp8_e5m2 إذا كان لديك CUDA

# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# نص الإدخال
input_text = "Hello, how are you?"

# ترميز النص
input_ids = tokenizer.encode(input_text, return_tensors="pt")

# إنشاء استدلال
with torch.no_grad():
    outputs = model.generate(input_ids, max_length=50)

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

print(generated_text)


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Hello, how are you?

I'm a little bit of a nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a


In [ ]:
39ث

In [1]:
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer

# تحميل النموذج والمحول
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
##kv_cache_dtype = torch.float16  # يمكنك استخدام fp8_e4m3fn أو fp8_e5m2 إذا كان لديك CUDA

# إعداد النموذج لاستخدام نوع البيانات المحدد
#model.config.kv_cache_dtype = kv_cache_dtype

# نص الإدخال
input_text = "Hello, how are you?"

# ترميز النص
input_ids = tokenizer.encode(input_text, return_tensors="pt")

# إنشاء استدلال
with torch.no_grad():
    outputs = model.generate(input_ids, max_length=50)

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

print(generated_text)


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Hello, how are you?

I'm a little bit of a nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a


In [ ]:
18ث

In [1]:
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer

# تحميل النموذج والمحول
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
kv_cache_dtype = torch.float16  # يمكنك استخدام fp8_e4m3fn أو fp8_e5m2 إذا كان لديك CUDA

# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# نص الإدخال
input_text = "Hello, how are you?"

# ترميز النص
input_ids = tokenizer.encode(input_text, return_tensors="pt")

# إنشاء استدلال
with torch.no_grad():
    outputs = model.generate(input_ids, max_length=50)

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

print(generated_text)


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Hello, how are you?

I'm a little bit of a nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a


In [ ]:
15ث

In [ ]:
kv_cache_dtype = torch.float16  # يمكنك استخدام fp8_e4m3fn أو fp8_e5m2 إذا كان لديك CUDA
kv_cache_dtype = torch.bfloat16
kv_cache_dtype = fp8_e4m3fn
kv_cache_dtype = fp8_e5m2





اشرح بالعربى معنى كل من kv_cache_dtype = torch.float16
kv_cache_dtype = torch.bfloat16 kv_cache_dtype = fp8_e4m3fn kv_cache_dtype = fp8_e5m2 وايهم اسرع

شرح معاني الأوامر:
جميع الأوامر التي ذكرتها تتعلق بتحديد نوع البيانات (dtype) المستخدمة في تخزين ذاكرة التخزين المؤقت للمفاتيح والقيم (key-value cache) في نماذج المحولات (transformer models) مثل GPT-2.

تأثير نوع البيانات: يؤثر نوع البيانات على دقة وسرعة النموذج. بشكل عام، استخدام أنواع بيانات ذات دقة أقل (مثل fp8) يزيد من السرعة ويقلل من استهلاك الذاكرة، ولكن قد يؤثر على دقة النموذج.

شرح كل أمر:

kv_cache_dtype = torch.float16: يحدد هذا الأمر نوع البيانات ليكون float16، وهو نوع بيانات بـ 16 بت يوفر توازنًا جيدًا بين السرعة والدقة. يُستخدم بشكل شائع في نماذج المحولات.

kv_cache_dtype = torch.bfloat16: يحدد هذا الأمر نوع البيانات ليكون bfloat16، وهو نوع بيانات بـ 16 بت آخر يوفر نطاقًا ديناميكيًا أوسع من float16، مما يجعله مناسبًا للحسابات التي تتضمن أرقامًا كبيرة جدًا أو صغيرة جدًا.

kv_cache_dtype = fp8_e4m3fn: يحدد هذا الأمر نوع البيانات ليكون fp8_e4m3fn، وهو نوع بيانات بـ 8 بت يوفر سرعة عالية ودقة أقل. يُستخدم بشكل أساسي في وحدات معالجة الرسومات (GPUs) الحديثة التي تدعم حسابات fp8.

kv_cache_dtype = fp8_e5m2: يحدد هذا الأمر نوع البيانات ليكون fp8_e5m2، وهو نوع بيانات بـ 8 بت آخر يوفر توازنًا مختلفًا بين السرعة والدقة مقارنة بـ fp8_e4m3fn. يُستخدم أيضًا في وحدات معالجة الرسومات الحديثة.

أيهما أسرع؟
بشكل عام، fp8_e4m3fn و fp8_e5m2 هما الأسرع نظرًا لأنهما يستخدمان نوع بيانات بـ 8 بت، مما يقلل من حجم البيانات المطلوب معالجتها. ومع ذلك، قد يؤدي ذلك إلى انخفاض طفيف في دقة النموذج.

torch.float16 يوفر توازنًا جيدًا بين السرعة والدقة، بينما torch.bfloat16 مناسب للحسابات التي تتضمن نطاقًا ديناميكيًا واسعًا.

ملاحظة: يعتمد اختيار نوع البيانات المناسب على متطلباتك الخاصة، مثل السرعة المطلوبة ودقة النموذج والموارد المتاحة.

In [2]:
!pip install GPUtil==1.4.0

  Preparing metadata (setup.py) ... done
  Created wheel for GPUtil: filename=GPUtil-1.4.0-py3-none-any.whl size=7392 sha256=bb2de7005de7bde6a825ee32fb8b08a7328933b55db0d058854d65f35c8b97ab
  Stored in directory: /root/.cache/pip/wheels/2b/4d/8f/55fb4f7b9b591891e8d3f72977c4ec6c7763b39c19f0861595
Successfully built GPUtil


In [1]:
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer
import time
import GPUtil

# تحميل النموذج والمحول
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
kv_cache_dtype = torch.float16  # يمكنك استخدام fp8_e4m3fn أو fp8_e5m2 إذا كان لديك CUDA

# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# نص الإدخال
input_text = "Hello, how are you?"

# ترميز النص
input_ids = tokenizer.encode(input_text, return_tensors="pt")

# بدء حساب وقت التنفيذ
start_time = time.time()

# إنشاء استدلال
with torch.no_grad():
    outputs = model.generate(input_ids, max_length=50)

# نهاية حساب وقت التنفيذ
end_time = time.time()

# حساب زمن التنفيذ
execution_time = end_time - start_time

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

# طباعة النص الناتج وزمن التنفيذ
print(generated_text)
print(f"زمن التنفيذ: {execution_time:.4f} ثانية")

# الحصول على معلومات حول استخدام GPU
GPUs = GPUtil.getGPUs()
for gpu in GPUs:
    print(f"GPU {gpu.id}:")
    print(f"  - اسم GPU: {gpu.name}")
    print(f"  - تحميل GPU: {gpu.load*100:.2f}%")
    print(f"  - ذاكرة مستخدمة: {gpu.memoryUsed:.2f} ميجابايت")
    print(f"  - ذاكرة إجمالية: {gpu.memoryTotal:.2f} ميجابايت")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Hello, how are you?

I'm a little bit of a nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a
زمن التنفيذ: 4.3061 ثانية
GPU 0:
  - اسم GPU: Tesla T4
  - تحميل GPU: 0.00%
  - ذاكرة مستخدمة: 0.00 ميجابايت
  - ذاكرة إجمالية: 15360.00 ميجابايت


In [1]:
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer
import time
import GPUtil

# تحميل النموذج والمحول
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
kv_cache_dtype = torch.bfloat16  # يمكنك استخدام fp8_e4m3fn أو fp8_e5m2 إذا كان لديك CUDA

# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# نص الإدخال
input_text = "Hello, how are you?"

# ترميز النص
input_ids = tokenizer.encode(input_text, return_tensors="pt")

# بدء حساب وقت التنفيذ
start_time = time.time()

# إنشاء استدلال
with torch.no_grad():
    outputs = model.generate(input_ids, max_length=50)

# نهاية حساب وقت التنفيذ
end_time = time.time()

# حساب زمن التنفيذ
execution_time = end_time - start_time

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

# طباعة النص الناتج وزمن التنفيذ
print(generated_text)
print(f"زمن التنفيذ: {execution_time:.4f} ثانية")

# الحصول على معلومات حول استخدام GPU
GPUs = GPUtil.getGPUs()
for gpu in GPUs:
    print(f"GPU {gpu.id}:")
    print(f"  - اسم GPU: {gpu.name}")
    print(f"  - تحميل GPU: {gpu.load*100:.2f}%")
    print(f"  - ذاكرة مستخدمة: {gpu.memoryUsed:.2f} ميجابايت")
    print(f"  - ذاكرة إجمالية: {gpu.memoryTotal:.2f} ميجابايت")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Hello, how are you?

I'm a little bit of a nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a
زمن التنفيذ: 3.1766 ثانية
GPU 0:
  - اسم GPU: Tesla T4
  - تحميل GPU: 0.00%
  - ذاكرة مستخدمة: 0.00 ميجابايت
  - ذاكرة إجمالية: 15360.00 ميجابايت


In [1]:
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer
import time
import GPUtil
from transformers.models.llama.modeling_llama import LlamaConfig
# تحميل النموذج والمحول
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
#kv_cache_dtype = fp8_e5m2  # يمكنك استخدام fp8_e4m3fn أو fp8_e5m2 إذا كان لديك CUDA
kv_cache_dtype = LlamaConfig.FP8_E5M2
# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# نص الإدخال
input_text = "Hello, how are you?"

# ترميز النص
input_ids = tokenizer.encode(input_text, return_tensors="pt")

# بدء حساب وقت التنفيذ
start_time = time.time()

# إنشاء استدلال
with torch.no_grad():
    outputs = model.generate(input_ids, max_length=50)

# نهاية حساب وقت التنفيذ
end_time = time.time()

# حساب زمن التنفيذ
execution_time = end_time - start_time

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

# طباعة النص الناتج وزمن التنفيذ
print(generated_text)
print(f"زمن التنفيذ: {execution_time:.4f} ثانية")

# الحصول على معلومات حول استخدام GPU
GPUs = GPUtil.getGPUs()
for gpu in GPUs:
    print(f"GPU {gpu.id}:")
    print(f"  - اسم GPU: {gpu.name}")
    print(f"  - تحميل GPU: {gpu.load*100:.2f}%")
    print(f"  - ذاكرة مستخدمة: {gpu.memoryUsed:.2f} ميجابايت")
    print(f"  - ذاكرة إجمالية: {gpu.memoryTotal:.2f} ميجابايت")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


AttributeError: type object 'LlamaConfig' has no attribute 'FP8_E5M2'

In [ ]:
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer, LlamaConfig
import time
import GPUtil

# ... (rest of the code)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
kv_cache_dtype = getattr(LlamaConfig, "FP8_E5M2")  # Accessing the attribute correctly

# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# ... (rest of the code)

In [ ]:
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer
import time
import GPUtil

# ... (rest of your code)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
kv_cache_dtype = "fp8_e5m2"

# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# ... (rest of your code)

In [1]:
#!pip install GPUtil==1.4.0
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer, LlamaConfig
import time
import GPUtil

# تحميل النموذج والمحول
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
try:
    # Trying to access FP8_E5M2 from LlamaConfig
    kv_cache_dtype = getattr(LlamaConfig, "FP8_E5M2")
except AttributeError:
    # If FP8_E5M2 is not found, fallback to torch.float16
    kv_cache_dtype = torch.float16
    print("FP8_E5M2 not found in LlamaConfig, using torch.float16 instead.")


# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# نص الإدخال
input_text = "Hello, how are you?"

# ترميز النص
input_ids = tokenizer.encode(input_text, return_tensors="pt")

# بدء حساب وقت التنفيذ
start_time = time.time()

# إنشاء استدلال
with torch.no_grad():
    outputs = model.generate(input_ids, max_length=50)

# نهاية حساب وقت التنفيذ
end_time = time.time()

# حساب زمن التنفيذ
execution_time = end_time - start_time

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

# طباعة النص الناتج وزمن التنفيذ
print(generated_text)
print(f"زمن التنفيذ: {execution_time:.4f} ثانية")

# الحصول على معلومات حول استخدام GPU
GPUs = GPUtil.getGPUs()
for gpu in GPUs:
    print(f"GPU {gpu.id}:")
    print(f"  - اسم GPU: {gpu.name}")
    print(f"  - تحميل GPU: {gpu.load*100:.2f}%")
    print(f"  - ذاكرة مستخدمة: {gpu.memoryUsed:.2f} ميجابايت")
    print(f"  - ذاكرة إجمالية: {gpu.memoryTotal:.2f} ميجابايت")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


FP8_E5M2 not found in LlamaConfig, using torch.float16 instead.
Hello, how are you?

I'm a little bit of a nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a
زمن التنفيذ: 2.2356 ثانية
GPU 0:
  - اسم GPU: Tesla T4
  - تحميل GPU: 0.00%
  - ذاكرة مستخدمة: 0.00 ميجابايت
  - ذاكرة إجمالية: 15360.00 ميجابايت


In [1]:
#!pip install GPUtil==1.4.0
import torch
from transformers import GPT2LMHeadModel, GPT2Tokenizer, LlamaConfig
import time
import GPUtil

# تحميل النموذج والمحول
model_name = "gpt2"
model = GPT2LMHeadModel.from_pretrained(model_name)
tokenizer = GPT2Tokenizer.from_pretrained(model_name)

# تحديد نوع البيانات للذاكرة المؤقتة للمفاتيح والقيم
try:
    # Trying to access FP8_E4M3FN from LlamaConfig
    kv_cache_dtype = getattr(LlamaConfig, "FP8_E4M3FN")  # Changed to FP8_E4M3FN
except AttributeError:
    # If FP8_E4M3FN is not found, fallback to torch.float16
    kv_cache_dtype = torch.float16
    print("FP8_E4M3FN not found in LlamaConfig, using torch.float16 instead.")

# إعداد النموذج لاستخدام نوع البيانات المحدد
model.config.kv_cache_dtype = kv_cache_dtype

# نص الإدخال
input_text = "Hello, how are you?"

# ترميز النص
input_ids = tokenizer.encode(input_text, return_tensors="pt")

# بدء حساب وقت التنفيذ
start_time = time.time()

# إنشاء استدلال
with torch.no_grad():
    outputs = model.generate(input_ids, max_length=50)

# نهاية حساب وقت التنفيذ
end_time = time.time()

# حساب زمن التنفيذ
execution_time = end_time - start_time

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

# طباعة النص الناتج وزمن التنفيذ
print(generated_text)
print(f"زمن التنفيذ: {execution_time:.4f} ثانية")

# الحصول على معلومات حول استخدام GPU
GPUs = GPUtil.getGPUs()
for gpu in GPUs:
    print(f"GPU {gpu.id}:")
    print(f"  - اسم GPU: {gpu.name}")
    print(f"  - تحميل GPU: {gpu.load*100:.2f}%")
    print(f"  - ذاكرة مستخدمة: {gpu.memoryUsed:.2f} ميجابايت")
    print(f"  - ذاكرة إجمالية: {gpu.memoryTotal:.2f} ميجابايت")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


FP8_E4M3FN not found in LlamaConfig, using torch.float16 instead.
Hello, how are you?

I'm a little bit of a nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a big nerd. I'm a
زمن التنفيذ: 2.2676 ثانية
GPU 0:
  - اسم GPU: Tesla T4
  - تحميل GPU: 0.00%
  - ذاكرة مستخدمة: 0.00 ميجابايت
  - ذاكرة إجمالية: 15360.00 ميجابايت


In [1]:
#!pip install GPUtil==1.4.0
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, LlamaConfig
import time
import GPUtil

# Function to load the model and tokenizer based on model_name
def load_model_and_tokenizer(model_name):
    model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto")
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    return model, tokenizer

# Model names to test
model_names = [
    "deepseek-ai/DeepSeek-R1-Distill-Qwen-7B",
    "gpt2",  # Add other model names here
    # ...
]

for model_name in model_names:
    print(f"Running inference for model: {model_name}")

    model, tokenizer = load_model_and_tokenizer(model_name)

    # Attempt to set kv_cache_dtype if supported by the model
    try:
        kv_cache_dtype = getattr(LlamaConfig, "FP8_E5M2")
        model.config.kv_cache_dtype = kv_cache_dtype
    except (AttributeError, torch.nn.modules.module.ModuleAttributeError):
        print(f"kv_cache_dtype not supported or found for {model_name}, skipping.")

    # Input text and generation
    input_text = "Hello, how are you?"
    input_ids = tokenizer.encode(input_text, return_tensors="pt")

    start_time = time.time()
    with torch.no_grad():
        outputs = model.generate(input_ids, max_length=50)
    end_time = time.time()

    # ... (Rest of the code for printing output and GPU info remains the same)



    # نهاية حساب وقت التنفيذ
#end_time = time.time()

# حساب زمن التنفيذ
execution_time = end_time - start_time

# فك ترميز النص الناتج
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

# طباعة النص الناتج وزمن التنفيذ
print(generated_text)
print(f"زمن التنفيذ: {execution_time:.4f} ثانية")

# الحصول على معلومات حول استخدام GPU
GPUs = GPUtil.getGPUs()
for gpu in GPUs:
    print(f"GPU {gpu.id}:")
    print(f"  - اسم GPU: {gpu.name}")
    print(f"  - تحميل GPU: {gpu.load*100:.2f}%")
    print(f"  - ذاكرة مستخدمة: {gpu.memoryUsed:.2f} ميجابايت")
    print(f"  - ذاكرة إجمالية: {gpu.memoryTotal:.2f} ميجابايت")

Running inference for model: deepseek-ai/DeepSeek-R1-Distill-Qwen-7B


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/181 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/3.06k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/7.03M [00:00<?, ?B/s]

AttributeError: module 'torch.nn.modules.module' has no attribute 'ModuleAttributeError'

In [1]:
#!pip install GPUtil==1.4.0  # Install GPUtil if not already installed
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, LlamaConfig
import time
import GPUtil
from torch.nn.modules import ModuleAttributeError  # Import for the error

def load_model_and_tokenizer(model_name):
    model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto")
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    return model, tokenizer

model_names = [
    "deepseek-ai/DeepSeek-R1-Distill-Qwen-7B",
    # Add other model names here
]

for model_name in model_names:
    print(f"Running inference for model: {model_name}")

    model, tokenizer = load_model_and_tokenizer(model_name)

try:
    kv_cache_dtype = getattr(LlamaConfig, "FP8_E5M2")
    model.config.kv_cache_dtype = kv_cache_dtype
except AttributeError:  # Catch only AttributeError
    print(f"kv_cache_dtype not supported or found for {model_name}, skipping.")
    input_text = "Hello, how are you?"
    input_ids = tokenizer.encode(input_text, return_tensors="pt")

    start_time = time.time()
    with torch.no_grad():
        outputs = model.generate(input_ids, max_length=50)
    end_time = time.time()

    execution_time = end_time - start_time
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    print(generated_text)
    print(f"Execution time: {execution_time:.4f} seconds")

    GPUs = GPUtil.getGPUs()
    for gpu in GPUs:
        print(f"GPU {gpu.id}:")
        print(f"  - Name: {gpu.name}")
        print(f"  - Load: {gpu.load * 100:.2f}%")
        print(f"  - Memory Used: {gpu.memoryUsed:.2f} MB")
        print(f"  - Total Memory: {gpu.memoryTotal:.2f} MB")

ImportError: cannot import name 'ModuleAttributeError' from 'torch.nn.modules' (/usr/local/lib/python3.11/dist-packages/torch/nn/modules/__init__.py)

In [1]:
#!pip install GPUtil==1.4.0  # Install GPUtil if not already installed
import torch
import torch.nn as nn
from transformers import AutoModelForCausalLM, AutoTokenizer, LlamaConfig
import time
import GPUtil

# Function to load the model and tokenizer based on model_name
def load_model_and_tokenizer(model_name):
    model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto")
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    return model, tokenizer

model_names = [
    "deepseek-ai/DeepSeek-R1-Distill-Qwen-7B",
    # Add other model names here
]

for model_name in model_names:
    print(f"Running inference for model: {model_name}")

    model, tokenizer = load_model_and_tokenizer(model_name)

    # Attempt to set kv_cache_dtype if supported by the model
    try:
        kv_cache_dtype = getattr(LlamaConfig, "FP8_E5M2")
        model.config.kv_cache_dtype = kv_cache_dtype
    except (AttributeError, torch.nn.modules.module.ModuleAttributeError): # Correctly access ModuleAttributeError
        print(f"kv_cache_dtype not supported or found for {model_name}, skipping.")

    # Input text and generation
    input_text = "Hello, how are you?"
    input_ids = tokenizer.encode(input_text, return_tensors="pt")

    start_time = time.time()
    with torch.no_grad():
        outputs = model.generate(input_ids, max_length=50)
    end_time = time.time()

    execution_time = end_time - start_time
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    print(generated_text)
    print(f"Execution time: {execution_time:.4f} seconds")

    GPUs = GPUtil.getGPUs()
    for gpu in GPUs:
        print(f"GPU {gpu.id}:")
        print(f"  - Name: {gpu.name}")
        print(f"  - Load: {gpu.load * 100:.2f}%")
        print(f"  - Memory Used: {gpu.memoryUsed:.2f} MB")
        print(f"  - Total Memory: {gpu.memoryTotal:.2f} MB")

Running inference for model: deepseek-ai/DeepSeek-R1-Distill-Qwen-7B


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

AttributeError: module 'torch.nn.modules.module' has no attribute 'ModuleAttributeError'

شغال gpu

In [ ]:
#!pip install GPUtil==1.4.0  # Install GPUtil if not already installed
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, LlamaConfig
import time
import GPUtil
import torch.nn as nn
# Function to load the model and tokenizer based on model_name
def load_model_and_tokenizer(model_name):
    model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto")
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    return model, tokenizer

model_names = [
    "deepseek-ai/DeepSeek-R1-Distill-Qwen-7B",
    # Add other model names here
]

for model_name in model_names:
    print(f"Running inference for model: {model_name}")

    model, tokenizer = load_model_and_tokenizer(model_name)

    # Attempt to set kv_cache_dtype if supported by the model
    # Check if the model is a Llama model before trying to set kv_cache_dtype
    if isinstance(model.config, LlamaConfig):  # Check if config is LlamaConfig
        try:
            kv_cache_dtype = getattr(LlamaConfig, "FP8_E5M2")
            model.config.kv_cache_dtype = kv_cache_dtype
        except AttributeError:
            print(f"kv_cache_dtype (FP8_E5M2) not found in LlamaConfig, skipping.")
    else:
        print(f"Model {model_name} is not a Llama model, skipping kv_cache_dtype setting.")

    # Input text and generation
    input_text = "Hello, how are you?"
    input_ids = tokenizer.encode(input_text, return_tensors="pt")

    start_time = time.time()
    with torch.no_grad():
        outputs = model.generate(input_ids, max_length=50)
    end_time = time.time()

    execution_time = end_time - start_time
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    print(generated_text)
    print(f"Execution time: {execution_time:.4f} seconds")

    GPUs = GPUtil.getGPUs()
    for gpu in GPUs:
        print(f"GPU {gpu.id}:")
        print(f"  - Name: {gpu.name}")
        print(f"  - Load: {gpu.load * 100:.2f}%")
        print(f"  - Memory Used: {gpu.memoryUsed:.2f} MB")
        print(f"  - Total Memory: {gpu.memoryTotal:.2f} MB")

In [1]:
#!pip install GPUtil==1.4.0  # Install GPUtil if not already installed
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, LlamaConfig
import time
import GPUtil

# Function to load the model and tokenizer based on model_name
def load_model_and_tokenizer(model_name):
    model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto")
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    return model, tokenizer

model_names = [
    "deepseek-ai/DeepSeek-R1-Distill-Qwen-7B",
    # Add other model names here
]

for model_name in model_names:
    print(f"Running inference for model: {model_name}")

    model, tokenizer = load_model_and_tokenizer(model_name)

    # Attempt to set kv_cache_dtype if supported by the model
    # Check if the model is a Llama model before trying to set kv_cache_dtype
    if isinstance(model.config, LlamaConfig):  # Check if config is LlamaConfig
        try:
            kv_cache_dtype = getattr(LlamaConfig, "FP8_E5M2")
            model.config.kv_cache_dtype = kv_cache_dtype
        except AttributeError:
            print(f"kv_cache_dtype (FP8_E5M2) not found in LlamaConfig, skipping.")
    else:
        print(f"Model {model_name} is not a Llama model, skipping kv_cache_dtype setting.")

    # Input text and generation
    input_text = "Hello, how are you?"
    input_ids = tokenizer.encode(input_text, return_tensors="pt")

    start_time = time.time()
    with torch.no_grad():
        outputs = model.generate(input_ids, max_length=50)
    end_time = time.time()

    execution_time = end_time - start_time
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    print(generated_text)
    print(f"Execution time: {execution_time:.4f} seconds")

    GPUs = GPUtil.getGPUs()
    for gpu in GPUs:
        print(f"GPU {gpu.id}:")
        print(f"  - Name: {gpu.name}")
        print(f"  - Load: {gpu.load * 100:.2f}%")
        print(f"  - Memory Used: {gpu.memoryUsed:.2f} MB")
        print(f"  - Total Memory: {gpu.memoryTotal:.2f} MB")

Running inference for model: deepseek-ai/DeepSeek-R1-Distill-Qwen-7B


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:151643 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Model deepseek-ai/DeepSeek-R1-Distill-Qwen-7B is not a Llama model, skipping kv_cache_dtype setting.


/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2134: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(


Hello, how are you? I have this problem where I need to find a positive integer solution for x and y in the equation: 1/x + 1/y = 1/4. I tried substituting some numbers, but I
Execution time: 1015.1227 seconds
GPU 0:
  - Name: Tesla T4
  - Load: 5.00%
  - Memory Used: 13317.00 MB
  - Total Memory: 15360.00 MB


Running inference for model: deepseek-ai/DeepSeek-R1-Distill-Qwen-7B
/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning:
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
Loading checkpoint shards: 100%
 2/2 [01:10<00:00, 34.23s/it]
WARNING:accelerate.big_modeling:Some parameters are on the meta device because they were offloaded to the cpu and disk.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:151643 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Model deepseek-ai/DeepSeek-R1-Distill-Qwen-7B is not a Llama model, skipping kv_cache_dtype setting.
/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2134: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(
Hello, how are you? I have this problem where I need to find a positive integer solution for x and y in the equation: 1/x + 1/y = 1/4. I tried substituting some numbers, but I
Execution time: 1015.1227 seconds
GPU 0:
  - Name: Tesla T4
  - Load: 5.00%
  - Memory Used: 13317.00 MB
  - Total Memory: 15360.00 MB

In [2]:
    input_text = "Who is Napoleon Bonaparte?"
    input_ids = tokenizer.encode(input_text, return_tensors="pt")

    start_time = time.time()
    with torch.no_grad():
        outputs = model.generate(input_ids, max_length=128)
    end_time = time.time()

    execution_time = end_time - start_time
    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

    print(generated_text)
    print(f"Execution time: {execution_time:.4f} seconds")

    GPUs = GPUtil.getGPUs()
    for gpu in GPUs:
        print(f"GPU {gpu.id}:")
        print(f"  - Name: {gpu.name}")
        print(f"  - Load: {gpu.load * 100:.2f}%")
        print(f"  - Memory Used: {gpu.memoryUsed:.2f} MB")
        print(f"  - Total Memory: {gpu.memoryTotal:.2f} MB")

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:151643 for open-end generation.
/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2134: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(


Who is Napoleon Bonaparte? What is the significance of his name?

Napoleon Bonaparte was a French military and political leader who is often referred to as the Emperor of the French. He led the French Revolution and later the Napoleonic Wars, achieving a brief unification of Europe. His rule was marked by both achievements and failures, and he is remembered with a complex legacy in history.

Napoleon's name, however, is not just a random combination of letters. It has a deeper significance tied to his life and legacy. The name "Napoleon" itself has a rich history, with roots in
Execution time: 3267.8593 seconds
GPU 0:
  - Name: Tesla T4
  - Load: 18.00%
  - Memory Used: 13327.00 MB
  - Total Memory: 15360.00 MB
